sql/core/src/main/scala/org/apache/spark/sql/execution/commands/parsers.scala

```diff
@@ -0,0 +1,420 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.commands
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.sql.{AnalysisException, SaveMode}
+import org.apache.spark.sql.catalyst.{CatalystQl, PlanParser, TableIdentifier}
+import org.apache.spark.sql.catalyst.analysis.UnresolvedRelation
+import org.apache.spark.sql.catalyst.expressions.{Ascending, Descending}
+import org.apache.spark.sql.catalyst.parser.{ASTNode, ParserConf, SimpleParserConf}
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, OneRowRelation}
+import org.apache.spark.sql.catalyst.plans.logical
+import org.apache.spark.sql.execution.commands._
+import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.types.StructType
+
+case class AlterTableCommandParser(base: CatalystQl) extends PlanParser {
+
+  def parsePartitionSpec(node: ASTNode): Option[Map[String, Option[String]]] = {
+    node match {
+      case Token("TOK_PARTSPEC", partitions) =>
+        val spec = partitions.map {
+          case Token("TOK_PARTVAL", ident :: constant :: Nil) =>
+            (unquoteString(cleanIdentifier(ident.text)),
+              Some(unquoteString(cleanIdentifier(constant.text))))
+          case Token("TOK_PARTVAL", ident :: Nil) =>
+            (unquoteString(cleanIdentifier(ident.text)), None)
+        }.toMap
+        Some(spec)
+      case _ => None
+    }
+  }
+
+  def extractTableProps(node: ASTNode): Map[String, Option[String]] = node match {
+    case Token("TOK_TABLEPROPERTIES", propsList) =>
+      propsList.flatMap {
+        case Token("TOK_TABLEPROPLIST", props) =>
+          props.map {
+            case Token("TOK_TABLEPROPERTY", key :: Token("TOK_NULL", Nil) :: Nil) =>
+              val k = unquoteString(cleanIdentifier(key.text))
+              (k, None)
+            case Token("TOK_TABLEPROPERTY", key :: value :: Nil) =>
+              val k = unquoteString(cleanIdentifier(key.text))
+              val v = unquoteString(cleanIdentifier(value.text))
+              (k, Some(v))
+          }
+      }.toMap
+  }
+
+  override def isDefinedAt(node: ASTNode): Boolean = node.text == "TOK_ALTERTABLE"
+
+  override def apply(v1: ASTNode): LogicalPlan = v1.children match {
+    case (tabName @ Token("TOK_TABNAME", _)) :: rest =>
+      val tableIdent: TableIdentifier = base.extractTableIdent(tabName)
+      val partitionSpec = base.getClauseOption("TOK_PARTSPEC", v1.children)
```

Comment:
Some alter table commands have partition spec defined. We should parse it at one place instead of doing this parsing at many places.